### ISSUE TITLE
About the structure of the Decoder

### ISSUE THREAD

**Issue by Zhaohuai-L** [Lijiaxin0111/Open-d4rt#16]
Hi, thanks for sharing the implementation of D4RT!
I noticed that the structure of the decoder in the code is somewhat different from the paper. In the paper, the original D4RT Decoder has an MLP module both before the Query and after the cross-attention, but I couldn't seem to locate these modules in the code.
Did I misunderstand something, or is there some other reason for this?

**The original design in the paper:**

<img width="1087" height="572" alt="Image" src="https://github.com/user-attachments/assets/face013f-63e1-4f01-8e31-9be3435c803d" />

**The design in the implementation:**
https://github.com/Lijiaxin0111/Open-d4rt/blob/bead824a40f2d719246469aed3acad101287a201/src/model/decoder.py#L43-L65

**Comment by Lijiaxin0111** [comment]
Hi, thanks for carefully checking the implementation and pointing this out!

You are right that the decoder structure in the current implementation is slightly different from the one described in the paper.

In the implementation, the MLP after the cross-attention is actually included inside each CrossAttentionBlock. Specifically, the self.ff module consists of the two linear layers:

https://github.com/Lijiaxin0111/Open-d4rt/blob/bead824a40f2d719246469aed3acad101287a201/src/model/decoder.py#L11-L40

so the post-cross-attention MLP is not missing.

However, you are also correct that the original D4RT decoder in the paper contains an additional MLP module before the query tokens are fed into the decoder block. In the current implementation, this query-side MLP projection/refinement is not explicitly implemented. This was an omission in the reproduction, and the current code directly uses the query tokens as the input of the first cross-attention block.

Thanks for pointing this out! I will update the implementation to make the decoder structure closer to the original paper by adding the missing query MLP module before the query.

**Comment by Zhaohuai-L** [comment]
> Hi, thanks for carefully checking the implementation and pointing this out!
> 
> You are right that the decoder structure in the current implementation is slightly different from the one described in the paper.
> 
> In the implementation, the MLP after the cross-attention is actually included inside each CrossAttentionBlock. Specifically, the self.ff module consists of the two linear layers:
> 
> [Open-d4rt/src/model/decoder.py](https://github.com/Lijiaxin0111/Open-d4rt/blob/bead824a40f2d719246469aed3acad101287a201/src/model/decoder.py#L11-L40)
> 
> Lines 11 to 40 in [bead824](/Lijiaxin0111/Open-d4rt/commit/bead824a40f2d719246469aed3acad101287a201)
> 
>  class CrossAttentionBlock(nn.Module): 
>      """Decoder block with cross-attention only (no query self-attention).""" 
>   
>      def __init__(self, hidden_dim: int, num_heads: int, mlp_ratio: float, dropout: float = 0.1) -> None: 
>          super().__init__() 
>          self.norm_q = nn.LayerNorm(hidden_dim) 
>          self.norm_kv = nn.LayerNorm(hidden_dim) 
>          self.cross_attn = nn.MultiheadAttention( 
>              embed_dim=hidden_dim, 
>              num_heads=num_heads, 
>              dropout=dropout, 
>              batch_first=True, 
>          ) 
>          ff_dim = int(math.ceil(hidden_dim * mlp_ratio)) 
>          self.norm_ff = nn.LayerNorm(hidden_dim) 
>          self.ff = nn.Sequential( 
>              nn.Linear(hidden_dim, ff_dim), 
>              nn.GELU(), 
>              nn.Dropout(dropout), 
>              nn.Linear(ff_dim, hidden_dim), 
>              nn.Dropout(dropout), 
>          ) 
>   
>      def forward(self, query_tokens: torch.Tensor, memory_tokens: torch.Tensor) -> torch.Tensor: 
>          q = self.norm_q(query_tokens) 
>          kv = self.norm_kv(memory_tokens) 
>          attn_out, _ = self.cross_attn(q, kv, kv, need_weights=False) 
>          x = query_tokens + attn_out 
>          x = x + self.ff(self.norm_ff(x)) 
>          return x 
> so the post-cross-attention MLP is not missing.
> 
> However, you are also correct that the original D4RT decoder in the paper contains an additional MLP module before the query tokens are fed into the decoder block. In the current implementation, this query-side MLP projection/refinement is not explicitly implemented. This was an omission in the reproduction, and the current code directly uses the query tokens as the input of the first cross-attention block.
> 
> Thanks for pointing this out! I will update the implementation to make the decoder structure closer to the original paper by adding the missing query MLP module before the query.

Hi Jiaxin, thanks for your reply! Looking forward to the upcoming implementation and checkpoints!
